plugins/clipboard/plugin.js

```diff
@@ -2539,6 +2583,227 @@
 			return undefined;
 		}
 	};
+
+	/**
+	 * Fallback dataTransfer object which is used together with {@link CKEDITOR.plugins.clipboard.dataTransfer}
+	 * for browsers supporting Clipboard API, but not supporting custom MIME types (Edge 16+).
+	 *
+	 * @since 4.8
+	 * @class CKEDITOR.plugins.clipboard.fallbackDataTransfer
+	 * @constructor Creates a class instance.
+	 * @param {Object} [nativeDataTransfer] A native data transfer object.
+	 * @param {String} [customDataFallbackType=text/html] A MIME type used for storing custom data types.
+	 */
+	CKEDITOR.plugins.clipboard.fallbackDataTransfer = function( nativeDataTransfer, customDataFallbackType ) {
+
+		/**
+		 * A native dataTransfer object.
+		 *
+		 * @private
+		 * @property {Object} _nativeDataTransfer
+		 */
+		this._nativeDataTransfer = nativeDataTransfer;
+
+		/**
+		 * A MIME type used for storing custom MIME types. Defaults to 'text/html'.
+		 *
+		 * @private
+		 * @property {String} _customDataFallbackType
+		 */
+		this._customDataFallbackType =  customDataFallbackType || 'text/html';
+	};
+
+	/**
+	 * True if the environment supports custom MIME types in `dataTransfer/cliboardData` `getData/setData` methods. Introduced
+	 * to distinguish browsers which supports only some whitelisted types (like 'text/html', 'application/xml'), but does not support
+	 * custom MIME types (like `cke/id`) like Edge 16+.
+	 * This property should not be accessed directly, use {@link CKEDITOR.plugins.clipboard.fallbackDataTransfer#isRequired}
+	 * method instead.
+	 *
+	 * @private
+	 * @static
+	 * @property {Boolean}
+	 */
+	CKEDITOR.plugins.clipboard.fallbackDataTransfer._isCustomMimeTypeSupported = null;
+
+	CKEDITOR.plugins.clipboard.fallbackDataTransfer.prototype = {
+		/**
+		 * Whether {@link CKEDITOR.plugins.clipboard.fallbackDataTransfer} should be used when operating on dataTransfer.
+		 * If true is returned, it means custom MIME types are not supported in the current browser
+		 * (see {@link CKEDITOR.plugins.clipboard.fallbackDataTransfer#_isCustomMimeTypeSupported}).
+		 *
+		 * @returns {Boolean}
+		 */
+		isRequired: function() {
+			// If there is no `dataTransfer` we cannot detect if fallback is needed.
+			// Method returns `false` so regular flow will be applied.
+			if ( !this._nativeDataTransfer ) {
+				return false;
+			}
+
+			if ( CKEDITOR.plugins.clipboard.fallbackDataTransfer._isCustomMimeTypeSupported === null ) {
+
+				CKEDITOR.plugins.clipboard.fallbackDataTransfer._isCustomMimeTypeSupported = true;
+
+				try {
+					this._nativeDataTransfer.setData( 'cke/mimetypetest', 'test value' );
```

Comment:
> That way catch could be simplified too.

Not sure what you meant here. We can do it like:
```
try {
	this._nativeDataTransfer.setData( testType, testValue );
	fallbackDataTransfer._isCustomMimeTypeSupported = this._nativeDataTransfer.getData( testType ) === testValue;
	this._nativeDataTransfer.setData( testType, '' );
} catch ( e ) {
	// Still needed as Edge will throw an error on first `setData`.
        if ( this._isUnsupportedMimeTypeError( e ) ) {
	    fallbackDataTransfer._isCustomMimeTypeSupported = false;
	}
}
```

or like:

```
try {
	this._nativeDataTransfer.setData( testType, testValue );
} catch ( e ) {}

fallbackDataTransfer._isCustomMimeTypeSupported = this._nativeDataTransfer.getData( testType ) === testValue;

try {
	this._nativeDataTransfer.setData( testType, '' );
} catch ( e ) {}
```

but it doesn't look good:D

WDYT @mlewand? Maybe you had something different on mind?

Reply:
One more thing - by checking the error code we narrow this to only Edge specific situation. If we will just check if data was set and there will be some other error thrown in other browser (well it shouldn't be, but not sure if we can be 100% sure) it will generate incorrect result (will still work with the fallback, but it is not optimal).

Reply:
@f1ames I was thinking more about:

```js
fallbackDataTransfer._isCustomMimeTypeSupported = false;

try {
	this._nativeDataTransfer.setData( testType, testValue );
	fallbackDataTransfer._isCustomMimeTypeSupported = this._nativeDataTransfer.getData( testType ) === testValue;
	this._nativeDataTransfer.setData( testType, '' );
} catch ( e ) {
}
```

BTW: `this._nativeDataTransfer.setData( testType, '' );` sets given type to an empty string, meaning that it will still be included in the clipboard, do we have a way to unset given data? Say `this._nativeDataTransfer.setData( testType, undefined );` or whatever? (Forgive me the lazyness for not checking it on my own now :P)

Reply:
Makes sense, I got fixed on initial `fallbackDataTransfer._isCustomMimeTypeSupported = true;` 😅 

We could use [clearData method](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/clearData) here. As it removes the given type from [types](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/types). However, [getData](https://developer.mozilla.org/en-US/docs/Web/API/DataTransfer/getData) works in the way (according to specs) that even for non-exisitng mime types it returns empty string.